clang/lib/CodeGen/CGExprConstant.cpp

```diff
@@ -1950,8 +1950,22 @@ ConstantLValueEmitter::tryEmitBase(const APValue::LValueBase &base) {
     if (D->hasAttr<WeakRefAttr>())
       return CGM.GetWeakRefReference(D).getPointer();
 
-    if (auto FD = dyn_cast<FunctionDecl>(D))
-      return CGM.GetAddrOfFunction(FD);
+    if (auto FD = dyn_cast<FunctionDecl>(D)) {
+      auto *C = CGM.GetAddrOfFunction(FD);
+
+      // we don't normally emit debug info for extern fns referenced via
+      // variable initialisers; BPF needs it since it generates BTF from
+      // debug info and bpftool demands BTF for every symbol linked
+      if (CGM.getTarget().getTriple().isBPF() && FD->getStorageClass() == SC_Extern) {
```

Comment:
There's an existing hook allowDebugInfoForExternalRef() for which targets want this.

You almost never want getStorageClass(); if you care about linkage, we have different methods for that.

Reply:
@efriedma-quic Thank you for insightful comments! As a newbie, I need more guidance.

It is pretty much clear that `allowDebugInfoForExternalRef` is a preferred way to check whether we should emit additional debug info. It is currently only overriden by BPF target and is being used for exact same purpose we have here.

I need help getting up to speed irt. `CompleteExternalDeclaration` proposal. Having grepped through the code, I can see that BPF faced the same challenge with `extern` variables. At the moment, there's a vector of `VariableDecl`s populated in `SemaDecl.cpp` and later used to attach debug info to variables.

A different mechanism is applied to discovering `extern` functions in need for debug info (prior to my patch, `extern` functions referenced through code were handled). This is accomplished by code in `CGExpr.cpp`. I am proposing a similar extension in `CGExprConstant.cpp`.

Would you like to have functions on `ExternalDeclarations` list instead?

What would be a good spot to get ahold of `FunctionDecl` to put it on `ExternalDeclarations` list?

Reply:
Looking at the code again, I guess the ultimate question is whether we want to emit debug info for all external functions/variables, or only functions/variables that are actually referenced.

If we want all functions/variables, we want to extend ExternalDeclarations to include them.  Maybe put the code in Sema::ActOnFunctionDeclarator, or something like that.

If you just want referenced functions/variables, probably the code should be in GetAddrOfFunction() (and something similar for variables).

Reply:
@mejedi @efriedma-quic Yes, we only want referenced functions/variables. The support for external variable has been implemented here https://github.com/llvm/llvm-project/commit/d77ae1552fc21a9f3877f3ed7e13d631f517c825. Thanks!

Reply:
@efriedma-quic @yonghong-song I ended up generalising `ExternalDeclarations` to include both `VarDecl`s and `FunctionDecl`s.
Since `GetAddrOfFunction` is invoked in so many different contexts, I found it difficult to emit debug info only for the subset of calls we care about.
Personally, I find it cleaner to collect external declarations on a list and handle them separately rather than plugging directly into codegen. Please let me know if I am missing something.
Concerning referenced/all debacle -
 * `Sema::ActOnEndOfTranslationUnit` checks whether a declaration is used (does it mean referenced?) before calling `Consumer.CompleteExternalDeclaration`;
 * plugging into codegen ensures that we notice any reference that made it into llvm bitcode BUT they could be removed by the optimiser later;
 * BPF backend ensures that only referenced entities make it to BTF.

It looks like both approaches are quite similar?

Reply:
Thanks @mejedi I think we do not need to worry about cases where external functions are in the code but removed by optimizer later as this should not that frequent. Your patch looks good to me, which follows what we did for VarDecl so adding FunctionDecl is a natural addition. BPF side of the change looks good to me too. @efriedma-quic could you take a look as well?

Reply:
Thanks @efriedma-quic if the clang believes the declaration might not be necessary due to optimizaiton, it is okay to remove it as in that case kernel/bpf does not need it either.